File: framework/wcf/Basic/Management/CircularTracing/CS/CircularTraceListener/CircularTraceListener.cs
//------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

using System;
using System.Diagnostics;
using System.Globalization;
using System.IO;

namespace Microsoft.Samples.ServiceModel
{

    public class CircularTraceListener : XmlWriterTraceListener
    {

        static CircularStream m_stream = null;
        bool MaxQuotaInitialized = false;
        const string FileQuotaAttribute = "maxFileSizeKB";
        const long DefaultMaxQuota = 1000;
        const string DefaultTraceFile = "E2ETraces.svclog";


        #region Member Functions

        private long MaxQuotaSize
        {
            //Get the MaxQuotaSize from configuration file
            //Set to Default Value if there are any problems

            get
            {
                long MaxFileQuota = 0;
                if (!this.MaxQuotaInitialized)
                {
                    try
                    {
                        string MaxQuotaOption = this.Attributes[CircularTraceListener.FileQuotaAttribute];
                        if (MaxQuotaOption == null)
                        {
                            MaxFileQuota = DefaultMaxQuota;
                        }
                        else
                        {
                            MaxFileQuota = int.Parse(MaxQuotaOption, CultureInfo.InvariantCulture);
                        }
                    }
                    catch (Exception)
                    {
                        MaxFileQuota = DefaultMaxQuota;
                    }
                    finally
                    {
                        this.MaxQuotaInitialized = true;
                    }
                }

                if (MaxFileQuota <= 0)
                {
                    MaxFileQuota = DefaultMaxQuota;
                }

                //MaxFileQuota is in KB in the configuration file, convert to bytes

                MaxFileQuota = MaxFileQuota * 1024;
                return MaxFileQuota;
            }
        }

        private void DetermineOverQuota()
        {

            //Set the MaxQuota on the stream if it hasn't been done

            if (!this.MaxQuotaInitialized)
            {
                m_stream.MaxQuotaSize = this.MaxQuotaSize;
            }

            //If we're past the Quota, flush, then switch files

            if (m_stream.IsOverQuota)
            {
                base.Flush();
                m_stream.SwitchFiles();
            }
        }

        #endregion               

        #region XmlWriterTraceListener Functions

        public CircularTraceListener(string file)
            : base(m_stream = new CircularStream(file))
        {
        }

        public CircularTraceListener()
            : base(m_stream = new CircularStream(DefaultTraceFile))
        {
        }

        protected override string[] GetSupportedAttributes()
        {
            return new string[] { CircularTraceListener.FileQuotaAttribute };
        }

        public override void TraceData(TraceEventCache eventCache, string source, TraceEventType eventType, int id, object data)
        {
            DetermineOverQuota();
            base.TraceData(eventCache, source, eventType, id, data);
        }

        public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id)
        {
            DetermineOverQuota();
            base.TraceEvent(eventCache, source, eventType, id);
        }

        public override void TraceData(TraceEventCache eventCache, string source, TraceEventType eventType, int id, params object[] data)
        {
            DetermineOverQuota();
            base.TraceData(eventCache, source, eventType, id, data);
        }

        public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id, string format, params object[] args)
        {
            DetermineOverQuota();
            base.TraceEvent(eventCache, source, eventType, id, format, args);
        }

        public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id, string message)
        {
            DetermineOverQuota();
            base.TraceEvent(eventCache, source, eventType, id, message);
        }

        public override void TraceTransfer(TraceEventCache eventCache, string source, int id, string message, Guid relatedActivityId)
        {
            DetermineOverQuota();
            base.TraceTransfer(eventCache, source, id, message, relatedActivityId);

        }

        #endregion

    }

    public class CircularStream : System.IO.Stream
    {
        private FileStream[] FStream = null;
        private String[] FPath = null;
        private long DataWritten = 0;
        private long FileQuota = 0;
        private int CurrentFile = 0;
        private string stringWritten = string.Empty;


        public CircularStream(string FileName)
        {
            //Handle all exceptions within this class, since tracing shouldn't crash a service

            //Add 00 and 01 to FileNames and open streams

            try
            {
                string filePath = Path.GetDirectoryName(FileName);
                string fileBase = Path.GetFileNameWithoutExtension(FileName);
                string fileExt = Path.GetExtension(FileName);

                if (string.IsNullOrEmpty(filePath))
                {
                    filePath = AppDomain.CurrentDomain.BaseDirectory;
                }

                FPath = new String[2];
                FPath[0] = Path.Combine(filePath, fileBase + "00" + fileExt);
                FPath[1] = Path.Combine(filePath, fileBase + "01" + fileExt);

                FStream = new FileStream[2];
                FStream[0] = new FileStream(FPath[0], FileMode.Create);
            }
            catch { }

        }

        public long MaxQuotaSize
        {
            get
            {
                return FileQuota;
            }
            set
            {
                FileQuota = value;
            }
        }

        public void SwitchFiles()
        {
            try
            {
                //Close current file, open next file (deleting its contents)

                DataWritten = 0;
                FStream[CurrentFile].Close();

                CurrentFile = (CurrentFile + 1) % 2;

                FStream[CurrentFile] = new FileStream(FPath[CurrentFile], FileMode.Create);
            }
            catch (Exception) { }
        }

        public bool IsOverQuota
        {
            get
            {
                return (DataWritten >= FileQuota);
            }

        }

        public override bool CanRead
        {
            get
            {
                try
                {
                    return FStream[CurrentFile].CanRead;
                }
                catch (Exception)
                {
                    return true;
                }
            }
        }

        public override bool CanSeek
        {
            get
            {
                try
                {
                    return FStream[CurrentFile].CanSeek;
                }
                catch (Exception)
                {
                    return false;
                }
            }
        }

        public override long Length
        {
            get
            {
                try
                {
                    return FStream[CurrentFile].Length;
                }
                catch (Exception)
                {
                    return -1;
                }
            }
        }

        public override long Position
        {
            get
            {
                try
                {
                    return FStream[CurrentFile].Position;
                }
                catch (Exception)
                {
                    return -1;
                }
            }
            set
            {
                try
                {
                    FStream[CurrentFile].Position = Position;
                }
                catch (Exception) { }
            }
        }

        public override bool CanWrite
        {
            get
            {
                try
                {
                    return FStream[CurrentFile].CanWrite;
                }
                catch (Exception)
                {
                    return true;
                }
            }
        }

        public override void Flush()
        {
            try
            {
                FStream[CurrentFile].Flush();
            }
            catch (Exception) { }
        }

        public override long Seek(long offset, SeekOrigin origin)
        {
            try
            {
                return FStream[CurrentFile].Seek(offset, origin);
            }
            catch (Exception)
            {
                return -1;
            }
        }

        public override void SetLength(long value)
        {
            try
            {
                FStream[CurrentFile].SetLength(value);
            }
            catch (Exception) { }
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            try
            {
                //Write to current file

                FStream[CurrentFile].Write(buffer, offset, count);
                DataWritten += count;

            }
            catch (Exception){ }
        }

        public override int Read(byte[] buffer, int offset, int count)
        {
            try
            {
                return FStream[CurrentFile].Read(buffer, offset, count);
            }
            catch (Exception)
            {
                return -1;
            }
        }

        public override void Close()
        {
            try
            {
                FStream[CurrentFile].Close();
            }
            catch (Exception) { }
        }


    }

}

File: framework/wcf/Basic/Management/CircularTracing/VB/CircularTraceListener/CircularTraceListener.vb
' Copyright (c) Microsoft Corporation.  All Rights Reserved.

Imports System
Imports System.Diagnostics
Imports System.IO
Imports System.Globalization

Namespace Microsoft.Samples.ServiceModel

    Public Class CircularTraceListener
        Inherits XmlWriterTraceListener

        Shared m_stream As CircularStream = Nothing
        Private MaxQuotaInitialized As Boolean = False
        Const FileQuotaAttribute As String = "maxFileSizeKB"
        Const DefaultMaxQuota As Long = 1000
        Const DefaultTraceFile As String = "E2ETraces.svclog"

#Region "Member Functions"

        Private ReadOnly Property MaxQuotaSize() As Long

            'Get the MaxQuotaSize from configuration file
            'Set to Default Value if there are any problems

            Get

                Dim MaxFileQuota As Long = 0
                If Not Me.MaxQuotaInitialized Then

                    Try

                        Dim MaxQuotaOption As String = Me.Attributes(CircularTraceListener.FileQuotaAttribute)
                        If MaxQuotaOption Is Nothing Then

                            MaxFileQuota = DefaultMaxQuota

                        Else

                            MaxFileQuota = Integer.Parse(MaxQuotaOption, CultureInfo.InvariantCulture)

                        End If

                    Catch generatedExceptionName As Exception

                        MaxFileQuota = DefaultMaxQuota

                    Finally

                        Me.MaxQuotaInitialized = True

                    End Try

                End If

                If MaxFileQuota <= 0 Then

                    MaxFileQuota = DefaultMaxQuota

                End If

                'MaxFileQuota is in KB in the configuration file, convert to bytes

                MaxFileQuota = MaxFileQuota * 1024
                Return MaxFileQuota

            End Get

        End Property

        Private Sub DetermineOverQuota()

            'Set the MaxQuota on the stream if it hasn't been done

            If Not Me.MaxQuotaInitialized Then

                m_stream.MaxQuotaSize = Me.MaxQuotaSize

            End If

            'If we're past the Quota, flush, then switch files

            If m_stream.IsOverQuota Then

                MyBase.Flush()
                m_stream.SwitchFiles()

            End If

        End Sub

#End Region

#Region "XmlWriterTraceListener Functions"

        Public Sub New(ByRef stream As CircularStream)
            MyBase.New(stream)

        End Sub

        Public Sub New(ByVal file As String)
            MyBase.New(file)
            m_stream = New CircularStream(file)
            Dim writer As StreamWriter = New StreamWriter(m_stream)
            MyBase.Writer = writer
        End Sub

        Protected Overloads Overrides Function GetSupportedAttributes() As String()

            Return New String() {CircularTraceListener.FileQuotaAttribute}

        End Function

        Public Overloads Overrides Sub TraceData(ByVal eventCache As TraceEventCache, ByVal source As String, ByVal eventType As TraceEventType, ByVal id As Integer, ByVal data As Object)

            DetermineOverQuota()
            MyBase.TraceData(eventCache, source, eventType, id, data)

        End Sub

        Public Overloads Overrides Sub TraceEvent(ByVal eventCache As TraceEventCache, ByVal source As String, ByVal eventType As TraceEventType, ByVal id As Integer)

            DetermineOverQuota()
            MyBase.TraceEvent(eventCache, source, eventType, id)

        End Sub

        Public Overloads Overrides Sub TraceData(ByVal eventCache As TraceEventCache, ByVal source As String, ByVal eventType As TraceEventType, ByVal id As Integer, ByVal ParamArray data As Object())

            DetermineOverQuota()
            MyBase.TraceData(eventCache, source, eventType, id, data)

        End Sub

        Public Overloads Overrides Sub TraceEvent(ByVal eventCache As TraceEventCache, ByVal source As String, ByVal eventType As TraceEventType, ByVal id As Integer, ByVal format As String, ByVal ParamArray args As Object())

            DetermineOverQuota()
            MyBase.TraceEvent(eventCache, source, eventType, id, format, args)

        End Sub

        Public Overloads Overrides Sub TraceEvent(ByVal eventCache As TraceEventCache, ByVal source As String, ByVal eventType As TraceEventType, ByVal id As Integer, ByVal message As String)

            DetermineOverQuota()
            MyBase.TraceEvent(eventCache, source, eventType, id, message)

        End Sub

        Public Overloads Overrides Sub TraceTransfer(ByVal eventCache As TraceEventCache, ByVal source As String, ByVal id As Integer, ByVal message As String, ByVal relatedActivityId As Guid)

            DetermineOverQuota()
            MyBase.TraceTransfer(eventCache, source, id, message, relatedActivityId)

        End Sub

#End Region

    End Class

    Public Class CircularStream
        Inherits System.IO.Stream

        Private FStream As FileStream() = Nothing
        Private FPath As String() = Nothing
        Private DataWritten As Long = 0
        Private FileQuota As Long = 0
        Private CurrentFile As Integer = 0
        Private stringWritten As String = String.Empty

        Public Sub New(ByVal FileName As String)

            'Handle all exceptions within this class, since tracing shouldn't crash a service

            'Add 00 and 01 to FileNames and open streams

            Try

                Dim filePath As String = Path.GetDirectoryName(FileName)
                Dim fileBase As String = Path.GetFileNameWithoutExtension(FileName)
                Dim fileExt As String = Path.GetExtension(FileName)

                If String.IsNullOrEmpty(filePath) Then

                    filePath = AppDomain.CurrentDomain.BaseDirectory

                End If

                FPath = New String(2) {}
                FPath(0) = Path.Combine(filePath, fileBase + "00" + fileExt)
                FPath(1) = Path.Combine(filePath, fileBase + "01" + fileExt)

                FStream = New FileStream(2) {}
                FStream(0) = New FileStream(FPath(0), FileMode.Create)

            Catch

            End Try

        End Sub

        Public Property MaxQuotaSize() As Long

            Get

                Return FileQuota

            End Get
            Set(ByVal value As Long)

                FileQuota = value

            End Set

        End Property

        Public Sub SwitchFiles()

            Try

                'Close current file, open next file (deleting its contents)

                DataWritten = 0
                FStream(CurrentFile).Close()

                CurrentFile = (CurrentFile + 1) Mod 2

                FStream(CurrentFile) = New FileStream(FPath(CurrentFile), FileMode.Create)

            Catch generatedExceptionName As Exception

            End Try

        End Sub

        Public ReadOnly Property IsOverQuota() As Boolean

            Get

                Return (DataWritten >= FileQuota)

            End Get

        End Property

        Public Overloads Overrides ReadOnly Property CanRead() As Boolean

            Get

                Try

                    Return FStream(CurrentFile).CanRead

                Catch generatedExceptionName As Exception

                    Return True

                End Try

            End Get

        End Property

        Public Overloads Overrides ReadOnly Property CanSeek() As Boolean

            Get

                Try

                    Return FStream(CurrentFile).CanSeek

                Catch generatedExceptionName As Exception

                    Return False

                End Try

            End Get

        End Property

        Public Overloads Overrides ReadOnly Property Length() As Long

            Get

                Try

                    Return FStream(CurrentFile).Length

                Catch generatedExceptionName As Exception

                    Return -1

                End Try

            End Get

        End Property

        Public Overloads Overrides Property Position() As Long

            Get

                Try

                    Return FStream(CurrentFile).Position

                Catch generatedExceptionName As Exception

                    Return -1

                End Try

            End Get
            Set(ByVal value As Long)

                Try

                    FStream(CurrentFile).Position = Position

                Catch generatedExceptionName As Exception

                End Try

            End Set

        End Property

        Public Overloads Overrides ReadOnly Property CanWrite() As Boolean

            Get

                Try

                    Return FStream(CurrentFile).CanWrite

                Catch generatedExceptionName As Exception

                    Return True

                End Try

            End Get

        End Property

        Public Overloads Overrides Sub Flush()

            Try

                FStream(CurrentFile).Flush()

            Catch generatedExceptionName As Exception

            End Try

        End Sub

        Public Overloads Overrides Function Seek(ByVal offset As Long, ByVal origin As SeekOrigin) As Long

            Try

                Return FStream(CurrentFile).Seek(offset, origin)

            Catch generatedExceptionName As Exception

                Return -1

            End Try

        End Function

        Public Overloads Overrides Sub SetLength(ByVal value As Long)

            Try

                FStream(CurrentFile).SetLength(value)

            Catch generatedExceptionName As Exception

            End Try

        End Sub

        Public Overloads Overrides Sub Write(ByVal buffer As Byte(), ByVal offset As Integer, ByVal count As Integer)

            Try

                'Write to current file
                FStream(CurrentFile).Write(buffer, offset, count)
                DataWritten += count

            Catch generatedExceptionName As Exception

            End Try

        End Sub

        Public Overloads Overrides Function Read(ByVal buffer As Byte(), ByVal offset As Integer, ByVal count As Integer) As Integer

            Try

                Return FStream(CurrentFile).Read(buffer, offset, count)

            Catch generatedExceptionName As Exception

                Return -1

            End Try

        End Function

        Public Overloads Overrides Sub Close()

            Try

                FStream(CurrentFile).Close()

            Catch generatedExceptionName As Exception

            End Try

        End Sub

    End Class

End Namespace
